driver-mapping/src/main/java/com/datastax/driver/mapping/ReflectionUtils.java

```diff
@@ -15,26 +15,179 @@
  */
 package com.datastax.driver.mapping;
 
-import java.lang.reflect.ParameterizedType;
-import java.lang.reflect.Type;
+import com.google.common.base.Throwables;
+
+import java.beans.BeanInfo;
+import java.beans.IntrospectionException;
+import java.beans.Introspector;
+import java.beans.PropertyDescriptor;
+import java.lang.annotation.Annotation;
+import java.lang.reflect.*;
+import java.util.HashMap;
+import java.util.Map;
 
 /**
  * Utility methods related to reflection.
  */
 class ReflectionUtils {
 
-    /**
-     * Gets a type argument from a parameterized type.
-     *
-     * @param pt   the parameterized type.
-     * @param arg  the index of the argument to retrieve.
-     * @param name the name of the element (field or method argument).
-     * @return the type argument.
-     */
-    static Class<?> getParam(ParameterizedType pt, int arg, String name) {
-        Type ft = pt.getActualTypeArguments()[arg];
-        if (!(ft instanceof Class))
-            throw new IllegalArgumentException(String.format("Cannot map parameter of class %s for %s", pt, name));
-        return (Class<?>) ft;
+    static <T> T newInstance(Class<T> clazz) {
+        Constructor<T> publicConstructor;
+        try {
+            publicConstructor = clazz.getConstructor();
+        } catch (NoSuchMethodException e) {
+            try {
+                // try private constructor
+                Constructor<T> privateConstructor = clazz.getDeclaredConstructor();
+                privateConstructor.setAccessible(true);
+                return privateConstructor.newInstance();
+            } catch (Exception e1) {
+                throw new IllegalStateException("Can't create an instance of " + clazz, e);
+            }
+        }
+        try {
+            return publicConstructor.newInstance();
+        } catch (Exception e) {
+            throw new IllegalStateException("Can't create an instance of " + clazz, e);
+        }
+    }
+
+    // for each key representing a property name,
+    // value[0] contains a Field object, value[1] contains a PropertyDescriptor object;
+    // they cannot be both null at the same time
+    static <T> Map<String, Object[]> scanFieldsAndProperties(Class<T> baseClass) {
+        Map<String, Object[]> fieldsAndProperties = new HashMap<String, Object[]>();
+        Map<String, Field> fields = scanFields(baseClass);
+        for (Map.Entry<String, Field> entry : fields.entrySet()) {
+            fieldsAndProperties.put(entry.getKey(), new Object[]{entry.getValue(), null});
+        }
+        Map<String, PropertyDescriptor> properties = scanProperties(baseClass);
+        for (Map.Entry<String, PropertyDescriptor> entry : properties.entrySet()) {
+            Object[] value = fieldsAndProperties.get(entry.getKey());
```

Comment:
What are the cases where you could have the multiple `PropertyDescriptor` with the same name?